From 61de2ff083add3b5af9821a71dddfae25919bc6b Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Thu, 18 Mar 2021 17:02:59 -0700
Subject: [PATCH 1785/1852] sound: xilinx: pcm: Assign device name per
 compatible string

commit b24bbb2246325c4c79673892c90340d5e7061964 from
https://github.com/Xilinx/linux-xlnx.git

Currently the device name is modified during probe. This causes an issue
when removing/unpopulating the device node because the sysfs files have
been created with the default device tree name instead of the modified
name from the probe function. So, if pcm is removed and reprobed we get
the following stack trace:
sysfs: cannot create duplicate filename '/bus/platform/devices/fd4a0000.
zynqmp-display:zynqmp_dp_snd_pcm0'

To avoid this, add new compatible strings and assign the device a name
according to the compatible string used. To maintain backwards
compatibility, use the old behavior if the old compatible string is
used.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
Acked-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xilinx-dp-pcm.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/sound/soc/xilinx/xilinx-dp-pcm.c b/sound/soc/xilinx/xilinx-dp-pcm.c
index 36c5afb50f4b..f32dc2c20e17 100644
--- a/sound/soc/xilinx/xilinx-dp-pcm.c
+++ b/sound/soc/xilinx/xilinx-dp-pcm.c
@@ -23,6 +23,9 @@
 #include <sound/pcm.h>
 #include <sound/soc.h>
 
+#define DP_PCM_NAME_0 "zynqmp_dp_snd_pcm0"
+#define DP_PCM_NAME_1 "zynqmp_dp_snd_pcm1"
+
 static const struct snd_pcm_hardware xilinx_pcm_hw = {
 	.info			= SNDRV_PCM_INFO_MMAP |
 				  SNDRV_PCM_INFO_MMAP_VALID |
@@ -46,7 +49,13 @@ static int xilinx_dp_pcm_probe(struct platform_device *pdev)
 {
 	int ret;
 
-	dev_set_name(&pdev->dev, pdev->dev.of_node->name);
+	if (of_device_is_compatible(pdev->dev.of_node, "xlnx,dp-snd-pcm0"))
+		dev_set_name(&pdev->dev, DP_PCM_NAME_0);
+	else if (of_device_is_compatible(pdev->dev.of_node, "xlnx,dp-snd-pcm1"))
+		dev_set_name(&pdev->dev, DP_PCM_NAME_1);
+	else
+		dev_set_name(&pdev->dev, pdev->dev.of_node->name);
+
 	pdev->name = dev_name(&pdev->dev);
 	ret = devm_snd_dmaengine_pcm_register(&pdev->dev,
 					      &xilinx_dmaengine_pcm_config, 0);
@@ -60,6 +69,8 @@ static int xilinx_dp_pcm_probe(struct platform_device *pdev)
 
 static const struct of_device_id xilinx_dp_pcm_of_match[] = {
 	{ .compatible = "xlnx,dp-snd-pcm", },
+	{ .compatible = "xlnx,dp-snd-pcm0", },
+	{ .compatible = "xlnx,dp-snd-pcm1", },
 	{ /* end of table */ },
 };
 MODULE_DEVICE_TABLE(of, xilinx_dp_pcm_of_match);
-- 
2.31.1

